src/base/emu-i386/simx86/codegen-sim.c

```diff
@@ -2960,6 +2956,13 @@ static __inline__ void SetCPU_WL(int m, signed char o, unsigned long v)
 unsigned int Sim_helper(unsigned int mem_ref, unsigned int data, int mode,
 			uint32_t *flags, unsigned int opc, unsigned int arg)
 {
+#ifdef X86_JIT
+	/* in the JIT we need the eip of translated code to get PC for a
+	   simulated page fault */
+	if (!CONFIG_CPUSIM)
+		currentIG = __builtin_return_address(0);
```

Comment:
Wow!
Why you suddenly need this?

Reply:
actually when I looked at it again, I don't, but the logic was wrong ever since I introduced Sim_helper.

One could have simulated page faults in Sim_helper, e.g. when ARPL or BOUND access memory, and they would longjmp and return the current interpreter cs:eip, which was wrong (this of course rarely happens). However Sim_helper already has exception check logic for other faults, and returns if so, so there is no need for longjmp any more at all in the JIT.

I think I should move Sim_helper() and the the sim_read/write functions it calls to a completely seperarate file then, like `helper.c` or `helper-sim.c`, as it's shared between JIT and sim? That way codegen-sim/fp87-sim are completely sim only.

Reply:
(the last commit here is a fixup, I should do a rebase if this passes)

Reply:
actually it's a bit more complicated since emu_pagefault_handler calls InvalidateNodeRange with the address of the compiled code; I'd need a version of InvalidateNodeRange that can take P0 instead of eip. Should be doable.